File: src/ops/assembly/adaptive/trace.rs
//! Adaptive-clearing trace record format and recorder adapter.
//!
//! Defines the per-step record serialised as MessagePack via rmp-serde.
//! The generic [`crate::trace::Tracer`] writes these records to the
//! self-contained trace file.
//!
//! [`TraceRecorder`] wraps an optional [`Tracer`] and exposes one-line
//! methods for each record type.  Runtime gating (via
//! [`AdaptiveClearingOptions::trace_path`]) means call sites in the
//! orchestrator are unconditional.

use serde::Serialize;

use crate::geo::algo::medial_axis::MedialAxis;
use crate::ops::assembly::adaptive::AdaptiveClearingOptions;
use crate::ops::container::Ops;
use crate::ops::cut::ClearedArea;
use crate::ops::cut::StepStatus;
use crate::trace::Tracer;
use crate::types::Point3D;

use super::tool::Tool;

// ── TraceKind ───────────────────────────────────────────────────────

/// Record kind byte values.
pub(crate) use crate::trace_types::TraceKind;

// ── Geometry / MAT records (emitted once at the start) ─────────────

/// Geometry parameters embedded in the trace file.
#[derive(Serialize)]
struct GeometryRecord {
    pub kind: &'static str,
    pub tool_radius: f64,
    pub boundary: Vec<(f64, f64)>,
    pub islands: Vec<Vec<(f64, f64)>>,
    pub seeds: Vec<Vec<(f64, f64)>>,
}

/// Lightweight serializable snapshot of the Medial Axis Transform.
///
/// Only the data needed for visualisation (nodes, clearances, edges, root)
/// — no LCA cache, no branches.
#[derive(Serialize)]
struct MatRecord {
    pub kind: &'static str,
    pub nodes: Vec<(f64, f64)>,
    pub clearances: Vec<f64>,
    pub edges: Vec<(u32, u32)>,
    pub root: u32,
}

impl From<&MedialAxis> for MatRecord {
    fn from(ma: &MedialAxis) -> Self {
        Self {
            kind: "mat",
            nodes: ma.nodes.iter().map(|n| (n.point.x, n.point.y)).collect(),
            clearances: ma.nodes.iter().map(|n| n.clearance).collect(),
            edges: ma
                .edges
                .iter()
                .map(|&(i, j)| (i as u32, j as u32))
                .collect(),
            root: ma.root as u32,
        }
    }
}

// ── TraceRecord ─────────────────────────────────────────────────────

/// Generic record header (same across all operations).
#[derive(Serialize, Clone, Debug)]
pub(super) struct TraceRecordHeader {
    pub kind: &'static str,
    pub status: u8,
    pub step_idx: u32,
    pub pos_x: f64,
    pub pos_y: f64,
    pub pos_z: f64,
    pub heading: f64,
    pub prev_x: f64,
    pub prev_y: f64,
    pub prev_z: f64,
    pub ops_len: u32,
}

/// Adaptive-clearing-specific payload nested inside every trace record.
#[derive(Serialize, Clone, Debug)]
pub(super) struct AdaptivePayload {
    pub iters: u32,
    pub smoothed_heading: f64,
    pub predicted_angle: f64,
    pub iteration_angle: f64,
    pub eng_angle: f64,
    pub eng_area: f64,
    pub eng_chord: f64,
    pub cut_area: f64,
    pub total_area: f64,
    pub remaining_area: f64,
    pub resume_source: u8,
    pub route_source: u8,
    /// Wall-hug points in resume order: current segment first, then
    /// previous segments from oldest to newest.
    pub wall_hug_points: Vec<(f64, f64)>,
    /// Per-segment point counts corresponding to `wall_hug_points`.
    /// First entry = current segment, remaining = previous segments
    /// from oldest to newest.
    pub wall_hug_segment_counts: Vec<u32>,
    /// Per-strategy outcome for the current resume attempt.
    /// Index 0-5 correspond to the strategy priority order:
    ///   WallHug, Segment, Mat, Frontier, Envelope, Island.
    /// Each byte: 0 = not tried, 1 = no candidate, 2 = blacklisted.
    pub resume_strategy_reasons: [u8; 6],
    /// Per-strategy detail code giving context for *why* the strategy
    /// returned None.  Parallel to `resume_strategy_reasons`.
    /// See `resume::DETAIL_*` constants for values.
    pub resume_strategy_details: [u8; 6],
    /// Per-strategy detail code for the last routing failure.
    /// Index 0-4 = Direct, Frontier, Mat, AStar, ZHop.
    /// 0 = success / not tried.  See `routing::ROUTE_*` constants.
    pub route_strategy_details: [u8; 5],
    /// Position of the last resume point candidate (routing target).
    pub resume_point_x: f64,
    pub resume_point_y: f64,
    pub resume_point_z: f64,
    /// Per-strategy candidate positions (x, y, z).  None entries are stored
    /// as (NaN, NaN, NaN).
    pub resume_candidate_points: [(f64, f64, f64); 6],
}

/// Per-step trace record, serialised as MessagePack.
///
/// Generic fields live at the top level; operation-specific data is in
/// the nested `payload` map so that a non-adaptive reader never sees
/// keys like ``eng_angle`` or ``wall_hug_points``.
#[derive(Serialize, Clone, Debug)]
pub(super) struct TraceRecord {
    #[serde(flatten)]
    pub header: TraceRecordHeader,
    pub payload: AdaptivePayload,
}

impl TraceRecord {
    /// Build a record with the common tool-state fields filled in from
    /// their source objects.  Kind-specific fields (iters, eng_*, etc.)
    /// default to 0 / 0.0.
    pub fn from_tool_state(
        kind: &'static str,
        status: StepStatus,
        step_idx: u32,
        tool: &Tool,
        cleared: &ClearedArea,
        prev_pos: Point3D,
        ops_len: u32,
    ) -> Self {
        let status: u8 = match status {
            StepStatus::Ok => 0,
            StepStatus::BoundaryHit => 1,
            StepStatus::LostEngagement => 2,
            StepStatus::NoConvergence => 3,
        };
        Self {
            header: TraceRecordHeader {
                kind,
                status,
                step_idx,
                pos_x: tool.pos.x,
                pos_y: tool.pos.y,
                pos_z: tool.pos.z,
                heading: tool.heading,
                prev_x: prev_pos.x,
                prev_y: prev_pos.y,
                prev_z: prev_pos.z,
                ops_len,
            },
            payload: AdaptivePayload {
                iters: 0,
                smoothed_heading: tool.smoothed_heading(),
                predicted_angle: tool.raw_predictor(),
                iteration_angle: 0.0,
                eng_angle: 0.0,
                eng_area: 0.0,
                eng_chord: 0.0,
                cut_area: 0.0,
                total_area: cleared.total_area(),
                remaining_area: cleared.remaining_area(),
                resume_source: 0,
                route_source: 0,
                wall_hug_points: Vec::new(),
                wall_hug_segment_counts: Vec::new(),
                resume_strategy_reasons: [0u8; 6],
                resume_strategy_details: [0u8; 6],
                route_strategy_details: [0u8; 5],
                resume_point_x: 0.0,
                resume_point_y: 0.0,
                resume_point_z: 0.0,
                resume_candidate_points: [(f64::NAN, f64::NAN, f64::NAN); 6],
            },
        }
    }
}

// ── TraceRecorder ───────────────────────────────────────────────────

/// Adapter that owns an optional [`Tracer`] and exposes one-line methods
/// for each record type.  When `tracer` is `None` all methods are
/// no-ops — call sites in the orchestrator are unconditional.
pub(super) struct TraceRecorder {
    tracer: Option<Tracer>,
    step_idx: u32,
}

impl TraceRecorder {
    /// Create a recorder, opening the trace file when
    /// `opts.trace_path` is `Some`.  Emits geometry and MAT records
    /// immediately after the header.
    pub fn new(
        opts: &AdaptiveClearingOptions,
        cleared: &ClearedArea,
        mat: Option<&MedialAxis>,
    ) -> Self {
        let tracer = match &opts.trace_path {
            Some(path) => match Tracer::open(path) {
                Ok(mut t) => {
                    let boundary: Vec<(f64, f64)> = opts
                        .pocket_boundary
                        .iter()
                        .map(|p| (p.x, p.y))
                        .collect();
                    let islands: Vec<Vec<(f64, f64)>> = opts
                        .islands
                        .iter()
                        .map(|poly| poly.iter().map(|p| (p.x, p.y)).collect())
                        .collect();
                    let seeds: Vec<Vec<(f64, f64)>> = cleared
                        .fragments()
                        .iter()
                        .map(|poly| poly.iter().map(|p| (p.x, p.y)).collect())
                        .collect();
                    t.write(&GeometryRecord {
                        kind: "geometry",
                        tool_radius: opts.radius,
                        boundary,
                        islands,
                        seeds,
                    });
                    if let Some(ma) = mat {
                        t.write(&MatRecord::from(ma));
                    }
                    Some(t)
                }
                Err(e) => {
                    eprintln!("trace: failed to open {:?}: {}", path, e);
                    None
                }
            },
            None => None,
        };

        Self {
            tracer,
            step_idx: 1,
        }
    }

    /// Record the initial tool state.
    pub fn record_init(
        &mut self,
        tool: &Tool,
        cleared: &ClearedArea,
        prev_pos: Point3D,
        ops_len: u32,
        wall_hug_points: &[(f64, f64)],
        wall_hug_segment_counts: &[u32],
    ) {
        if let Some(ref mut tr) = self.tracer {
            let mut rec = TraceRecord::from_tool_state(
                "init",
                StepStatus::Ok,
                0,
                tool,
                cleared,
                prev_pos,
                ops_len,
            );
            rec.payload.wall_hug_points = wall_hug_points.to_vec();
            rec.payload.wall_hug_segment_counts =
                wall_hug_segment_counts.to_vec();
            tr.write(&rec);
        }
    }

    /// Record a cut step.  Increments the internal step index.
    #[allow(clippy::too_many_arguments)]
    pub fn record_cut(
        &mut self,
        status: StepStatus,
        tool: &Tool,
        cleared: &ClearedArea,
        prev_pos: Point3D,
        ops_len: u32,
        iters: u32,
        iteration_angle: f64,
        eng_angle: f64,
        eng_area: f64,
        eng_chord: f64,
        cut_area: f64,
        wall_hug_points: &[(f64, f64)],
        wall_hug_segment_counts: &[u32],
    ) {
        if let Some(ref mut tr) = self.tracer {
            let mut rec = TraceRecord::from_tool_state(
                "cut",
                status,
                self.step_idx,
                tool,
                cleared,
                prev_pos,
                ops_len,
            );
            rec.payload.iters = iters;
            rec.payload.iteration_angle = iteration_angle;
            rec.payload.eng_angle = eng_angle;
            rec.payload.eng_area = eng_area;
            rec.payload.eng_chord = eng_chord;
            rec.payload.cut_area = cut_area;
            rec.payload.wall_hug_points = wall_hug_points.to_vec();
            rec.payload.wall_hug_segment_counts =
                wall_hug_segment_counts.to_vec();
            tr.write(&rec);
        }
        self.step_idx += 1;
    }

    /// Record a resume event (stall or stuck).  Increments the
    /// internal step index.
    #[allow(clippy::too_many_arguments)]
    pub fn record_resume(
        &mut self,
        kind: TraceKind,
        status: StepStatus,
        resume_source: u8,
        route_source: u8,
        tool: &Tool,
        cleared: &ClearedArea,
        prev_pos: Point3D,
        ops_len: u32,
        reasons: &[u8; 6],
        details: &[u8; 6],
        route_details: &[u8; 5],
        rp: Point3D,
        candidate_pts: &[(f64, f64, f64); 6],
        wall_hug_points: &[(f64, f64)],
        wall_hug_segment_counts: &[u32],
    ) {
        if let Some(ref mut tr) = self.tracer {
            let kind_str = match kind {
                TraceKind::ResumeStall => "resume_stall",
                TraceKind::ResumeStuck => "resume_stuck",
                _ => "resume",
            };
            let mut rec = TraceRecord::from_tool_state(
                kind_str,
                status,
                self.step_idx,
                tool,
                cleared,
                prev_pos,
                ops_len,
            );
            rec.payload.resume_source = resume_source;
            rec.payload.route_source = route_source;
            rec.payload.resume_strategy_reasons = *reasons;
            rec.payload.resume_strategy_details = *details;
            rec.payload.route_strategy_details = *route_details;
            rec.payload.resume_point_x = rp.x;
            rec.payload.resume_point_y = rp.y;
            rec.payload.resume_point_z = rp.z;
            rec.payload.resume_candidate_points = *candidate_pts;
            rec.payload.wall_hug_points = wall_hug_points.to_vec();
            rec.payload.wall_hug_segment_counts =
                wall_hug_segment_counts.to_vec();
            tr.write(&rec);
        }
        self.step_idx += 1;
    }

    /// Record an exit event.
    #[allow(clippy::too_many_arguments)]
    pub fn record_exit(
        &mut self,
        status: StepStatus,
        tool: &Tool,
        cleared: &ClearedArea,
        prev_pos: Point3D,
        ops_len: u32,
        reasons: &[u8; 6],
        details: &[u8; 6],
        route_details: &[u8; 5],
        rp: Point3D,
        candidate_pts: &[(f64, f64, f64); 6],
        wall_hug_points: &[(f64, f64)],
        wall_hug_segment_counts: &[u32],
    ) {
        if let Some(ref mut tr) = self.tracer {
            let mut rec = TraceRecord::from_tool_state(
                "exit",
                status,
                self.step_idx,
                tool,
                cleared,
                prev_pos,
                ops_len,
            );
            rec.payload.resume_strategy_reasons = *reasons;
            rec.payload.resume_strategy_details = *details;
            rec.payload.route_strategy_details = *route_details;
            rec.payload.resume_point_x = rp.x;
            rec.payload.resume_point_y = rp.y;
            rec.payload.resume_point_z = rp.z;
            rec.payload.resume_candidate_points = *candidate_pts;
            rec.payload.wall_hug_points = wall_hug_points.to_vec();
            rec.payload.wall_hug_segment_counts =
                wall_hug_segment_counts.to_vec();
            tr.write(&rec);
        }
    }

    /// Finalise the trace file.
    /// Write the toolpath block and finalise the trace file.
    pub fn finish(self, _ops: &Ops) {
        if let Some(mut t) = self.tracer {
            let _ = t.finish();
        }
    }
}

// ── Helpers (called from the recorder) ──────────────────────────────

File: python/raygeo/cli/trace.py
import struct

import msgpack

# ── Trace file format ────────────────────────────────────────────

TRACE_HEADER_SIZE = 12  # magic(4) + version(4) + count(4)
TRACE_MAGIC = b"ADPT"

KIND_NAMES = {
    0: "init",
    1: "cut",
    2: "resume_stall",
    3: "resume_stuck",
    4: "exit",
}
STATUS_NAMES = {
    0: "Ok",
    1: "BoundaryHit",
    2: "LostEngagement",
    3: "NoConvergence",
}

RESUME_SOURCE_NAMES = {
    0: "none",
    1: "wall_hug",
    2: "segment",
    3: "mat",
    4: "frontier",
    5: "island",
    6: "envelope",
}

ROUTE_SOURCE_NAMES = {
    0: "none",
    1: "direct",
    2: "frontier",
    3: "mat",
    4: "astar",
}

ROUTE_DETAIL_LABELS = {
    0: ".",
    1: "sweep_collide",
    2: "no_obstacles",
    3: "no_frontier",
    4: "offset_empty",
    5: "diff_polygons",
    6: "too_few_verts",
    7: "same_vertex",
    8: "sweep_collide",
    9: "no_axis",
    10: "no_cleared",
    11: "no_path",
    12: "sweep_collide",
    13: "no_obstacles",
    14: "no_free_space",
    15: "astar_failed",
    16: "too_few_waypoints",
}


class TraceRecord:
    """One per-step record from the trace file (msgpack dict wrapper)."""

    __slots__ = (
        "kind",
        "status",
        "step_idx",
        "iters",
        "pos_x",
        "pos_y",
        "heading",
        "smoothed_heading",
        "predicted_angle",
        "iteration_angle",
        "eng_angle",
        "eng_area",
        "eng_chord",
        "cut_area",
        "total_area",
        "remaining_area",
        "prev_x",
        "prev_y",
        "ops_len",
        "resume_source",
        "route_source",
        "wall_hug_points",
        "wall_hug_segment_counts",
        "resume_strategy_reasons",
        "resume_strategy_details",
        "route_strategy_details",
        "resume_point_x",
        "resume_point_y",
        "resume_candidate_points",
    )

    def __init__(self, d):
        for k, v in d.items():
            if k in self.__slots__:
                setattr(self, k, v)
        if not hasattr(self, "resume_strategy_details"):
            self.resume_strategy_details = [0] * 6
        if not hasattr(self, "route_strategy_details"):
            self.route_strategy_details = [0] * 4
        if not hasattr(self, "resume_point_x"):
            self.resume_point_x = 0.0
            self.resume_point_y = 0.0

    @property
    def wall_hug_count(self):
        return len(self.wall_hug_points)


class TraceGeometry:
    """Pocket geometry embedded in a trace file."""

    __slots__ = ("tool_radius", "boundary", "islands", "seeds")

    def __init__(self, tool_radius, boundary, islands, seeds):
        self.tool_radius = tool_radius
        self.boundary = boundary
        self.islands = islands
        self.seeds = seeds


class TraceFile:
    """Binary trace reader with random access to records.

    Geometry, seeds, toolpath, and per-step records are all embedded in
    a single self-contained file.  Records are length-prefixed MessagePack
    blobs.
    """

    def __init__(self, path):
        with open(path, "rb") as f:
            magic = f.read(4)
            if magic != TRACE_MAGIC:
                raise ValueError(f"bad magic: {magic}")
            f.read(4)  # reserved
            self.count = struct.unpack("<I", f.read(4))[0]
            self.geometry = self._read_geometry(f)
            self._read_mat(f)
            self.toolpath = self._read_toolpath(f)
            self._records = []
            for _ in range(self.count):
                rec_len = struct.unpack("<I", f.read(4))[0]
                rec_bytes = f.read(rec_len)
                rec_dict = msgpack.unpackb(rec_bytes)
                self._records.append(TraceRecord(rec_dict))

    def _read_geometry(self, f):
        tool_radius = struct.unpack("<d", f.read(8))[0]
        boundary = self._read_polygon(f)
        n_islands = struct.unpack("<I", f.read(4))[0]
        islands = [self._read_polygon(f) for _ in range(n_islands)]
        n_seeds = struct.unpack("<I", f.read(4))[0]
        seeds = [self._read_polygon(f) for _ in range(n_seeds)]
        return TraceGeometry(tool_radius, boundary, islands, seeds)

    @staticmethod
    def _read_polygon(f):
        n = struct.unpack("<I", f.read(4))[0]
        pts = []
        for _ in range(n):
            x = struct.unpack("<d", f.read(8))[0]
            y = struct.unpack("<d", f.read(8))[0]
            pts.append((x, y))
        return pts

    @staticmethod
    def _read_toolpath(f):
        n = struct.unpack("<I", f.read(4))[0]
        pts = []
        for _ in range(n):
            x = struct.unpack("<d", f.read(8))[0]
            y = struct.unpack("<d", f.read(8))[0]
            is_travel = f.read(1) != b"\x00"
            pts.append((x, y, bool(is_travel)))
        return pts

    def _read_mat(self, f):
        present = struct.unpack("<B", f.read(1))[0]
        if present:
            n_nodes = struct.unpack("<I", f.read(4))[0]
            self.mat_nodes = []
            self.mat_clearances = []
            for _ in range(n_nodes):
                x = struct.unpack("<d", f.read(8))[0]
                y = struct.unpack("<d", f.read(8))[0]
                c = struct.unpack("<d", f.read(8))[0]
                self.mat_nodes.append((x, y))
                self.mat_clearances.append(c)
            n_edges = struct.unpack("<I", f.read(4))[0]
            self.mat_edges = []
            for _ in range(n_edges):
                i = struct.unpack("<I", f.read(4))[0]
                j = struct.unpack("<I", f.read(4))[0]
                self.mat_edges.append((i, j))
            self.mat_root = struct.unpack("<I", f.read(4))[0]
        else:
            self.mat_nodes = []
            self.mat_clearances = []
            self.mat_edges = []
            self.mat_root = 0

    def __len__(self):
        return len(self._records)

    def __getitem__(self, idx):
        if idx < 0 or idx >= len(self._records):
            raise IndexError(idx)
        return self._records[idx]
